File: services/rails-modular/app/controllers/api/sessions_controller.rb
# frozen_string_literal: true

module Api
  # Controller がやってよいのは 3 つだけ:
  #   1. 入力の検証
  #   2. パッケージの公開 API 呼び出し
  #   3. 応答の組み立て（例外 → HTTP ステータスの変換）
  #
  # **触ってよいのは Identity::Api だけ。**
  # User モデルや SignInUseCase を直接呼ぶと packwerk が落とす。
  class SessionsController < ApplicationController
    def create
      email = params.require(:email)
      password = params.require(:password)

      result = Identity::Api.sign_in(
        email: email,
        password: password,
        ip: request.remote_ip,
        user_agent: request.user_agent
      )

      render json: {
        access_token: result[:tokens].access_token,
        id_token: result[:tokens].id_token,
        refresh_token: result[:tokens].refresh_token,
        expires_in: result[:tokens].expires_in,
        user: result[:user].to_h
      }
    rescue Identity::Api::AuthenticationFailed => e
      # 業務の語彙（認証失敗）を HTTP の語彙（401）へ翻訳するのはここ
      render json: { detail: e.message }, status: :unauthorized
    rescue ActionController::ParameterMissing => e
      render json: { detail: "#{e.param} は必須です" }, status: :bad_request
    end

    def show
      # require_authentication! を通さないと、未認証時に
      # AuthenticationFailed がそのまま漏れて 500 になる（実際に踏んだ）。
      return unless authenticated?

      render json: current_user_view.to_h
    end
  end
end

File: services/cakephp-modular/src/Controller/Api/SessionsController.php
<?php
declare(strict_types=1);

namespace App\Controller\Api;

use App\Controller\AppController;
use Identity\Exception\AuthenticationFailedException;
use Identity\Public\IdentityApi;

/**
 * Controller がやってよいのは 3 つだけ:
 *   1. 入力の検証
 *   2. モジュールの公開 API 呼び出し
 *   3. 例外 → HTTP ステータスの変換
 *
 * **触ってよいのは Identity\Public\IdentityApi だけ。**
 * UsersTable や SignInUseCase を直接呼ぶと deptrac が落とす。
 */
class SessionsController extends AppController
{
    public function create(): void
    {
        $email = (string)$this->request->getData('email');
        $password = (string)$this->request->getData('password');

        if ($email === '' || strlen($password) < 8) {
            $this->respond(['detail' => 'メールアドレスとパスワード(8文字以上)は必須です'], 400);

            return;
        }

        try {
            $result = (new IdentityApi())->signIn(
                $email,
                $password,
                $this->request->clientIp(),
                $this->request->getHeaderLine('User-Agent')
            );
        } catch (AuthenticationFailedException $e) {
            // 業務の語彙（認証失敗）を HTTP の語彙（401）へ翻訳するのはここ
            $this->respond(['detail' => $e->getMessage()], 401);

            return;
        }

        $this->respond([
            'access_token' => $result['tokens']['access_token'],
            'id_token' => $result['tokens']['id_token'],
            'refresh_token' => $result['tokens']['refresh_token'],
            'expires_in' => $result['tokens']['expires_in'],
            'user' => $result['user'],
        ], 200);
    }

    public function me(): void
    {
        $user = $this->requireAuthentication();
        if ($user === null) {
            return;
        }

        $this->respond($user, 200);
    }
}
